File: app/pages/repos/_repo_id/models/_model_id/prediction/props.ts
import { PinwheelInfo } from '@tangramhq/pinwheel'
import { InputTable, Prediction } from 'common/predict'
import { ModelLayoutInfo } from 'layouts/model_layout'

export type Column = UnknownField | NumberField | EnumField | TextField

export type Props = {
	inner: Inner
	modelLayoutInfo: ModelLayoutInfo
	pinwheelInfo: PinwheelInfo
}

export enum InnerType {
	PredictionForm = 'prediction_form',
	PredictionResult = 'prediction_result',
}

export type Inner =
	| {
			type: InnerType.PredictionForm
			value: PredictionForm
	  }
	| {
			type: InnerType.PredictionResult
			value: PredictionResult
	  }

export type PredictionForm = {
	form: Form
}

export type Form = { fields: Column[] }

export type PredictionResult = {
	inputTable: InputTable
	prediction: Prediction
}

export enum FieldType {
	Unknown = 'unknown',
	Number = 'number',
	Enum = 'enum',
	Text = 'text',
}

export type UnknownField = {
	name: string
	type: FieldType.Unknown
	value: string | null
}

export type NumberField = {
	max: number
	min: number
	name: string
	p25: number
	p50: number
	p75: number
	type: FieldType.Number
	value: string | null
}

export type EnumField = {
	histogram: Array<[string, number]>
	name: string
	options: string[]
	type: FieldType.Enum
	value: string | null
}

export type TextField = {
	name: string
	type: FieldType.Text
	value: string | null
}

File: app/pages/repos/_repo_id/models/_model_id/prediction/props.rs
use crate::{common::predict::PredictionResult, layouts::model_layout::ModelLayoutInfo};

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Props {
	pub id: String,
	pub model_layout_info: ModelLayoutInfo,
	pub inner: Inner,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "type", content = "value")]
pub enum Inner {
	PredictionForm(PredictionForm),
	PredictionResult(PredictionResult),
}

#[derive(serde::Serialize)]
pub struct PredictionForm {
	pub form: PredictForm,
}

#[derive(serde::Serialize)]
pub struct PredictForm {
	pub fields: Vec<Column>,
}

#[derive(serde::Serialize)]
#[serde(tag = "type")]
pub enum Column {
	#[serde(rename = "unknown")]
	Unknown(Unknown),
	#[serde(rename = "number")]
	Number(Number),
	#[serde(rename = "enum")]
	Enum(Enum),
	#[serde(rename = "text")]
	Text(Text),
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Unknown {
	pub name: String,
	pub value: String,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Number {
	pub name: String,
	pub max: f32,
	pub min: f32,
	pub p25: f32,
	pub p50: f32,
	pub p75: f32,
	pub value: String,
}

#[derive(serde::Serialize)]
pub struct Enum {
	pub name: String,
	pub options: Vec<String>,
	pub value: String,
	pub histogram: Vec<(String, u64)>,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Text {
	pub name: String,
	pub value: String,
}
